dynamiqs/solvers/solver.py

```diff
@@ -89,29 +93,42 @@ def _run(self):
     def run_nograd(self):
         pass
 
+    def next_tsave(self) -> float:
+        if self.tsave_counter == len(self.tsave):
+            return inf
+        return self.tsave[self.tsave_counter].item()
+
+    def next_tmeas(self) -> float:
+        if self.tmeas_counter == len(self.tmeas):
+            return inf
+        return self.tmeas[self.tmeas_counter].item()
+
     def next_tstop(self) -> float:
         return self.tstop[self.tstop_counter].item()
 
     def save(self, y: Tensor):
-        self._save(y)
-        self.tstop_counter += 1
-
-    def _save(self, y: Tensor):
-        if self.tsave_mask[self.tstop_counter]:
+        if self.next_tstop() == self.next_tsave():
```

Comment:
I'm scared of this line and the equivalent one for `tmeas`, with numerical precision and the such in case `tsave` or `tmeas` are modified by anything in the code. I would propose instead to have something like:
```python
def merge_unique_sorted_arrays(X, Y):
    # Combine the arrays and find the unique elements (which are also sorted)
    Z = torch.unique(torch.cat((X, Y)))

    # Check which elements of Z are in X and Y to create the indicator arrays
    from_X = torch.isin(Z, X)
    from_Y = torch.isin(Z, Y)

    return Z, from_X, from_Y

# Test
X = torch.tensor([1, 3, 5, 7])
Y = torch.tensor([2, 3, 6, 8, 9])
Z, from_X, from_Y = merge_unique_sorted_arrays(X, Y)
print("Z:", Z.numpy())
print("From X:", from_X.numpy())
print("From Y:", from_Y.numpy())
# Z: [1 2 3 5 6 7 8 9]
# From X: [ True False  True  True False  True False False]
# From Y: [False  True  True False  True False  True  True]
```
which is called when defining `tstop`. And then one can just check the arrays `from_tsave` and `from_tmeas` to increment `self.tsave_counter` and `self.tmeas_counter`.

Reply:
I agree, this is what was done before, don't know why I changed this part.